### DISCUSSION TITLE
Crazyflie firmware commander issues

### DISCUSSION THREAD

**Discussion by Jasperdez** [bitcraze/discussions#671] [Q&A]
Good evening,

If have been working on a thesis in which i make use of the Crazyflie to measure light levels of LED's. One of the functions i am using is commanderSetSetpoint() which I want to use to make my Crazyflie fly to certain coordinates. However the function is not working as expected. For setpoint(0,0,0) it instantly flips on its back and for any value for the z coordinate it just flies in to the ceiling no matter wat I put in. I have set the modes for x,y,z  and yaw to modeAbs so it should be using absolute positioning.

I am assuming I am missing something here so maybe someone could help me with this.

```

static void Hover(setpoint_t *setpoint, uint8_t type, float x,float,y,float z,float yaw)
{

  setpoint->mode.x = modeAbs;
  setpoint->mode.y = modeAbs;
  setpoint->mode.z = modeAbs;

  setpoint->position.x = x;
  setpoint->position.y = y;
  setpoint->position.z = z;

  setpoint->mode.yaw = modeAbs;

  setpoint->attitude.yaw = yaw;

commanderSetSetpoint(setpoint,3);
}
```

**Comment by ataffanel** [comment]
Hi,

Your code looks correct in principle, however your question is missing a very important information: what positioning system are you using and is the position estimation working correctly?

If you are using one of the internal positioning system like Lighthouse, Loco Positioning or the flow deck, you can test it from the Crazyflie client using the command based flight control in the flight tab.

If you are using an external system like a Mocap, it might be a bit more complex and it will greatly depends of how you did setup the system.

**Reply by Jasperdez** [reply]
Thanks for the reply,

I was using the Lighthouse but due to certain circumstances I was told to realize it without any external positioning system which a fellow thesis student told me possible while a bit unstable. I have tried using the complementary and kalman estimators but don't notice much difference. Would it be possible to setup the crazyflie in such a way that it could fly with the internal sensors (albeit a bit unstable) or is this not feasable?

**Reply by ataffanel** [reply]
If you are only using internal sensors, it explains the behavior you are seeing: using the Kalman filter with only internal sensor will work well for a couple of seconds, after that the estimated velocity will drift too much which will make the estimated position drift very far very fast. Flying autonomously on internal sensor only is only viable during very short times like passing though occlusion blocking the lighthouse rays. For longer flight, you need some kind of extra sensor.
